bfvmm/src/hve/arch/intel_x64/vmexit/control_register.cpp

```diff
@@ -306,19 +330,56 @@ void
 control_register_handler::enable_wrcr4_exiting(
     vmcs_n::value_type mask)
 {
-    using namespace vmcs_n;
-    mask |= m_vcpu->global_state()->ia32_vmx_cr4_fixed0;
+    vmcs_n::cr4_guest_host_mask::set(
+        mask | m_vcpu->global_state()->ia32_vmx_cr4_fixed0
+    );
+}
+
+// -----------------------------------------------------------------------------
+// Execute Functions
+// -----------------------------------------------------------------------------
+
+void
+control_register_handler::execute_wrcr0(
+    gsl::not_null<vcpu *> vcpu)
+{
+    emulate_rdgpr(vcpu);
+    vcpu->set_gr2(vcpu->cr0());
+    vcpu->set_cr0(vcpu->gr1());
+}
 
-    cr4_guest_host_mask::set(mask);
-    cr4_read_shadow::set(guest_cr4::get());
+void
+control_register_handler::execute_rdcr3(
+    gsl::not_null<vcpu *> vcpu)
+{
+    vcpu->set_gr1(vcpu->cr3());
+    emulate_wrgpr(vcpu);
+}
+
+void
+control_register_handler::execute_wrcr3(
+    gsl::not_null<vcpu *> vcpu)
+{
+    emulate_rdgpr(vcpu);
+    vcpu->set_gr2(vcpu->cr3());
+    vcpu->set_cr3(vcpu->gr1());
+}
+
+void
+control_register_handler::execute_wrcr4(
+    gsl::not_null<vcpu *> vcpu)
+{
+    emulate_rdgpr(vcpu);
+    vcpu->set_gr2(vcpu->cr4());
+    vcpu->set_cr4(vcpu->gr1());
 }
 
 // -----------------------------------------------------------------------------
 // Handlers
 // -----------------------------------------------------------------------------
 
 bool
-control_register_handler::handle(vcpu_t vcpu)
+control_register_handler::handle(vcpu *vcpu)
```

Comment:
Why are some args wrapped with gsl::not_null but not others?

Reply:
All APIs that are "inputs" have gsl::not_null when the contract expects that. All "output" functions, which is mostly the hanlders do not as the user can assume that the APIs will not provide a null pointer. Technically... gsl::not_null should be used for output as this dictates "ensures" and not "expects", but in a previous converstation you me and Connor agreed that this was relatively pedantic, and made the delegates too verbose.

Reply:
Ok sgtm I just couldn't remember they were for inputs